
# Introductory Data Science Tools: Unix and Python

## GAIN program Summer 2023    
**Prof**: Dr. Thomas Parchman; SFB 209; tparchman@unr.edu   
**Co-instructor**: Trevor Faske; tfaske@nevada.unr.edu \
**Workshop hours**: 9:00-11:00 AM, August 20, 21, 


## Workshop Description
Modern biology, and other fields of science, are increasingly shaped by data sets that are orders of magnitude larger than many scientists have traditionally been trained to work with. For example, major recent advances in DNA sequencing technology have created the ability to affordably generate data spanning billions of DNA sequences in an extremely short period of time. Similar leaps in data acquisition technology are transforming other scientific disciplines, including but not limited to geography, communications, economics, chemistry, and physics. Spreadsheet software and graphical user interface statistical analysis packages (e.g., Excel, Statistica, JMP) are useless for this now common scale of data. The ability to manipulate, process, and analyze large data sets with basic programming and data science skills should accelerate the research productivity and success of graduate students in this era. 

This short module will introduce students to basic computational tools, focusing on Unix and Python, to support data proficiency. We will start at the most basic level, assuming zero or limited previous experience, and aim to build enough familiarity for students to feel equipped and motivated to learn more on their own. Our overall goal is to provide an entry point for students to gain further expertise in simple programming and efficient manipulation and analysis of large-scale data sets. As our time together is extremely limited, I aim to introduce you to linux and python, attempting to describe why writing code in these languages is useful. I have attempted to put together fairly easy to follow modules written in markdown so that you have introductory resources available for getting started with Linux and Python. The continued development of proficiency with these and similar skillsets are guaranteed to improve the efficacy and quality pf graduate research and productivity.

## Student Learning Outcomes
By the end of the module, students should have learned enough to feel enabled and motivated to learn more Unix and Python. Thus, this module will serve to introduce students to programming, to prepare them for more in depth courses, and most importantly to provide a foundation for ongoing independent learning. 

Specific Student Learning Outcomes:

- Students will be able to operate in a Unix computing environment, and will understand the basic use of Unix computing clusters for research.
     
- Students will be able to write basic programs in Python in order to efficiently manipulate and work with large scale data.

- Students will have enough exposure to freely available resources for Unix and Python to continue independently learning programming skills.

- Students will be able to use basic Unix and Python skills to manipulate large data sets, and to conduct basic analyses of genome level DNA sequencing data.

## Required Materials

- **Computer with Unix operating system** Students with
  Mac computers already have machines running Unix and
  are ready to go. Students without Mac computers will have the option of checking out a Mac laptop for the semester, or will need to figure out how to install Linux or a Linux emulator on their computer. 

- **Installed text editor with syntax recognition** Students should install a text editor that will recognize syntax from code written for Unix, Python, Perl, etc. I suggest BBedit (for mac users), Visual Studio Code, or Sublime. All are free and easy to locate, download, and install.

- **Supplemental primers, readings and assignments** are provided on the [workshop github page](https://github.com/tparchman/GAIN_2023).

### Suggested resources for more in depth learning
* **Practical computing for biologists** Haddock, S.H.D. and Dunn, C.W., 2011. Sunderland, MA, USA: Sinauer Associates. The book is very useful for both Unix and Python.

# Python documentation and other useful resources

[Python documentation](https://www.python.org/doc/)

[Python for Biologists](https://pythonforbiologists.com/introduction)

## Workshop Format
We will meet from 9:00-11:00 each weekday morning from August 6th to August 12th. At the beginning of each session, we will introduce new concepts and material that will form the basis of the markdown primers I provide for introductory exercises. We will cover questions or discussion, and then students will spend at least half of each period working through the primers independently or in small groups. Given limited time, I at least hope that these primers may serve a somewhat useful introduction for those wanting to learn more.

## Workshop Material Organization

The material for each day of the workshop will be organized in separate directories on [workshop github page](https://github.com/tparchman/GAIN_2023). Each of these directories will contain the slides that we will use to introduce material, a primer covering example Unix and Python code along with explanations, and a worksheet of programming practice exercises. There are also general directories on the repository with supplementary resources for Unix and Python, including cheat sheets, tutorials, and recommended resources for learning more.

### Downloading material from github using `git`

While you can download indidvidual files from github using your preferred web browser, you can also use the UNIX command to access github as well. Using git commands can get complicated very quickly, it is a very useful skill to have for reproducibility, tracking changes, and collaboration. We do not go over git in this course but there are many tutorials online (http://swcarpentry.github.io/git-novice/). 

For this course, downloading individual files might suffice. But if you would like to download the entire repo, you can do so through the command line using the below command:

**hint:** make a directory somewhere on your computer for this workshop. Run below command in that directory.

```
git clone https://github.com/tparchman/GAIN_2023
```

## Workshop Schedule
*Tentative Workshop Schedule. All contents are subject to change.

| Date    |  Topic          |  Assignment |
| --------- | ---------------| -------------| 
| Aug. 20  | Intro, Unix    | Unix assignments  | 
| Aug. 21  |	Python I | python1_practice_scripts.md |
| Aug. 21  |	Python II	   | python2_practice_scripts.md |
| ------ | ---------------| --------------------------------| 
|

